Add validation support for WSI  khr display extension

jon-lunarg opened the pull request:
Mark, Tobin, and Ian:
Can you review these commits for adding VK_KHR_display extension support to object_tracker, unique_objects  and swapchain?  The swapchain validation is pretty basic, there are probably more checks that could be added later.

Petros:
I squashed your series of commits from PR #355 into one commit that is in this PR.  Then I finished out the support for the extension (primarily in object_tracker and unique_objects).  Please review.  I think as far as object_tracker and unique_object layers they should completely support this extension now.

Commit:
Added WSI KHR_display support to validation

Add support for this extension to object_tracker and swapchain layers.

Change-Id: Ifdf6095958a9d9d01720fde92e0bbd0a8a9a3216

jon-lunarg committed:
layers: Finish add support to validation for VK_KHR_display extension

Swapchain, object_tracker and unique_objects now intercept and perform
validation checks for this extension.
Change-Id: I2e78d1dc516103de165ef83b99665eaad05b73d9

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -129,11 +129,26 @@ static void createInstanceRegisterExtensions(const VkInstanceCreateInfo *pCreate
+    pDisp->CreateDisplayModeKHR =
+        (PFN_vkCreateDisplayModeKHR)gpa(instance, "vkCreateDisplayModeKHR");
+    pDisp->GetDisplayPlaneCapabilitiesKHR =
+        (PFN_vkGetDisplayPlaneCapabilitiesKHR)gpa(instance, "vkGetDisplayPlaneCapabilitiesKHR");
```
Question: are the commands/functions from VK_KHR_display supposed to be statically exported from the loader, like the "core WSI extensions" (e.g. VK_KHR_surface)?  I honestly don't remember.  We should check, in case they need to be protected by an enabled-check.

ianelliottus replied:
This should be protected by #ifdef.  That's because not all platforms will support the VK_KHR_display[_swapchain] extensions.

This comment can effectively be "ditto'd" throughout this pull request.

ianelliottus replied:
I believe that I'll be filing a Khronos-private bug about whether the VK_KHR_display[_swapchain] extension will be statically exported.  The language in the spec was adopted before these extensions were added.  Overtime, I can see some other extensions being created that may not be supported across Android, Linux, and/or Windows.  Thus, we might need some caveats about which WSI extensions will have their commands/functions statically exported.

ianelliottus replied:
I did a little research and filed a private, spec issue (380) to address this.  At the time we discussed this, I believe we only had what I'll call the "core WSI extensions" (VK_KHR_{surface|swapchain}), which are fully supported on Android, Linux, and Windows.  Since then, we've added the 6 "platform extensions", which probably will only be supported on one OS each, and the 2 "display extensions", which will never be supported on Android.  The issue is to change the spec language a bit to clarify that only "relevant extensions" will have their functions statically exported by a given loader.

jon-lunarg replied:
Hi Ian:
Thanks for the code review!  I am not sure how the issue of weather the loader exports the KHR_display extension  entry points or not effects this code review.  The loader exports or doesn't export the entry points independently of any layer's decision to intercept  the extension entry points.  The above code is just initializing the layer's dispatch table for these functions.  So if the layer intercepts these functions  it can call down the chain by having it's dispatch table pointers initialized.

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -129,11 +129,26 @@ static void createInstanceRegisterExtensions(const VkInstanceCreateInfo *pCreate
     // was enabled for it:
     my_data->instanceMap[instance].instance = instance;
     my_data->instanceMap[instance].surfaceExtensionEnabled = false;
+    my_data->instanceMap[instance].displayExtensionEnabled = false;
```
This should be protected by #ifdef.  That's because not all platforms will support the VK_KHR_display[_swapchain] extensions.

This comment can effectively be "ditto'd" throughout this pull request.

jon-lunarg replied:
There is no ifdef in vulkan.h that pertains to this extension VK_KHR_display. Perhaps that is an oversight but I don't think so.  I think, the reason the VK_KHR_xlib_surface, VK_KHR_win32_surface, etc. extensions are protected by an ifdef in vulkan.h is because  these extensions # include platform specific headers (such as #include <X11/Xlib.h>) that will not compile on foreign platforms.  The VK_KHR_display extension doesn't # include any headers so it can compile on ANY platform even though it may not be supported on various platforms at run time by the driver.  Thus, if we ifdef this code we would have to invent a #define and the only advantage would be the layer library size would be smaller for platforms that don't support this extension. I think it is better to always compile this extension and let the app determine at run time if this extension is supported or not on the platform.

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+    if (!pPropertyCount) {
+        skipCall |= LOG_ERROR_NULL_POINTER(VK_DEBUG_REPORT_OBJECT_TYPE_INSTANCE_EXT, pPhysicalDevice->pInstance, "pPropertyCount");
+    }
+    lock.unlock();
```
Please look at the implementation of GetPhysicalDeviceSurfaceFormatsKHR() in this file, for the "else" case (i.e. if pPropertyCount is non-NULL).

jon-lunarg replied:
got it!

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetPhysicalDeviceDisplayPropertiesKHR(physicalDevice, pPropertyCount, pProperties);
```
if ((result == VK_SUCCESS) && ...) {
    // Record the result ...
}

jon-lunarg replied:
okay

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+
+    if (!pPropertyCount) {
+        skipCall |= LOG_ERROR_NULL_POINTER(VK_DEBUG_REPORT_OBJECT_TYPE_INSTANCE_EXT, pPhysicalDevice->pInstance, "pPropertyCount");
+    }
```
Please look at the implementation of GetPhysicalDeviceSurfaceFormatsKHR() in this file, for the "else" case (i.e. if pPropertyCount is non-NULL).

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+        if (!pPhysicalDevice->gotDisplayPlanePropertyCount)
+        {
+            pPhysicalDevice->displayPlanePropertyCount = *pPropertyCount;
+            pPhysicalDevice->gotDisplayPlanePropertyCount = true;
```
Should should be recording the VkDisplayKHR handles, for future validation against.

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetDisplayPlaneSupportedDisplaysKHR(physicalDevice, planeIndex, pDisplayCount, pDisplays);
+
```
Assuming VK_SUCCESS, you should should be recording the VkDisplayKHR handles, for future validation against.

jon-lunarg replied:
okay

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+}
+
+VKAPI_ATTR VkResult VKAPI_CALL 
+GetDisplayPlaneSupportedDisplaysKHR(VkPhysicalDevice physicalDevice, uint32_t planeIndex, uint32_t* pDisplayCount, VkDisplayKHR* pDisplays) {
```
Ditto some of the other comments from other queries

ianelliottus replied:
Also, please look at the "Valid Usage" sections of the spec for things to validate against, which come from other queries.

jon-lunarg replied:
okay

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+}
+
+VKAPI_ATTR VkResult VKAPI_CALL 
+GetDisplayModePropertiesKHR(VkPhysicalDevice physicalDevice, VkDisplayKHR display, uint32_t* pPropertyCount, VkDisplayModePropertiesKHR* pProperties) {
```
Ditto previous comments (which I won't keep repeating:-)

ianelliottus commented on layers/swapchain.cpp:
```diff
@@ -901,6 +920,292 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
+                              SWAPCHAIN_EXT_NOT_ENABLED_BUT_USED,
+                              "%s() called even though the %s extension was not enabled for this VkInstance.", __FUNCTION__,
+                              VK_KHR_DISPLAY_EXTENSION_NAME);
+    }
```
Please validate against previous queries (e.g. per Valid Usage).

jon-lunarg replied:
okay

ianelliottus commented on layers/swapchain.h:
```diff
@@ -161,6 +164,9 @@ struct SwpInstance {
     // Set to true if VK_KHR_SURFACE_EXTENSION_NAME was enabled for this VkInstance:
     bool surfaceExtensionEnabled;
 
+    // Set to true if VK_KHR_DISPLAY_EXTENSION_NAME was enabled for this VkInstance:
```
Please protect with #ifdef.

ianelliottus commented on layers/unique_objects.h:
```diff
@@ -81,6 +83,16 @@ static void createInstanceRegisterExtensions(const VkInstanceCreateInfo *pCreate
     pDisp->GetPhysicalDeviceSurfacePresentModesKHR =
         (PFN_vkGetPhysicalDeviceSurfacePresentModesKHR)gpa(instance, "vkGetPhysicalDeviceSurfacePresentModesKHR");
+
+    // KHR_display
```
Ditto on #ifdef.

ianelliottus commented on vk-layer-generate.py:
```diff
@@ -54,7 +54,14 @@ def proto_is_global(proto):
         "CreateAndroidSurfaceKHR",
         "CreateWin32SurfaceKHR",
-        "GetPhysicalDeviceWin32PresentationSupportKHR"
+        "GetPhysicalDeviceWin32PresentationSupportKHR",
```
Need some way of #ifdef'ing the display extension functions

ianelliottus commented:
-1.  Hi @jon-lunarg and @pbantolas.  I did a review (stopped after a bit).  This looks like a really good start, with a few issues I outlined in comments.  Best wishes!

jon-lunarg committed:
layers: Fix GetDisplayModePropertiesKHR unique_objects code

Add the available DisplayModes to the map for use by CreateDisplayPlaneSurfaceKHR

Change-Id: I94754259069a844bb516d3cf9d4e130473dbc2e3

jon-lunarg committed:
layers: FIx object_tracker displayKHR and displayModeKHR creation

For displayModeKHR  there are two separate Vulkan commands that can
create these objects.
For displayKHR there is one command that creates the object:
GetPhysicalDeviceDisplayPropertiesKHR.  GetDisplayPlaneSupportedDisplaysKHR
doesn't create new objects but returns existing objects.  The existing code
had the creation in  GetDisplayPlaneSupportedDisplaysKHR rather than in
GetPhysicalDeviceDisplayPropertiesKHR.

Change-Id: I63df7c41a0d17acd8b06ebb1a5742ac032159a09

jon-lunarg commented:
Petros:

I pushed two new commits to this branch. One that fixed the issue you pointed out in unique_objects. The other commit fixes object_tracker. 

As for your comment:
"In a similar fashion, the object_tracker layer requires that vkGetDisplayPlaneSupportedDisplaysKHR is called for any subsequent VkDisplayKHR validation to pass. Do you think create_display_khr should be introduced to other display-related functions, or should vkGetDisplayPlaneSupportedDisplaysKHR always be called first?"

My closer reading of the spec indicates that GetPhysicalDeviceDisplayPropertiesKHR actually creates the DisplayKHR objects so it must be called first. GetDisplayPlaneSupportedDisplaysKHR returns DisplayKHR objects that have already been created. So I skipped validating these output
parameters  nor putting them in the map.  Do you think this is correct?

ashleysmithgpu commented:
Hi, I believe that vkGetPhysicalDeviceDisplayPropertiesKHR need an explicit_ version too, like vkGetDisplayModePropertiesKHR:

```
    if (result == VK_SUCCESS && pProperties)
    {   
        for (uint32_t idx0=0; idx0<*pPropertyCount; ++idx0) {
            std::lock_guard<std::mutex> lock(global_lock);
            
            uint64_t unique_id = global_unique_id++;
            my_map_data->unique_id_mapping[unique_id] = reinterpret_cast<uint64_t &>(local_pProperties[idx0].display);
            pProperties[idx0].display = reinterpret_cast<VkDisplayKHR&>(unique_id);
            pProperties[idx0].displayName = local_pProperties[idx0].displayName;
            pProperties[idx0].physicalDimensions = local_pProperties[idx0].physicalDimensions;
            pProperties[idx0].physicalResolution = local_pProperties[idx0].physicalResolution;
            pProperties[idx0].supportedTransforms = local_pProperties[idx0].supportedTransforms;
            pProperties[idx0].planeReorderPossible = local_pProperties[idx0].planeReorderPossible;
            pProperties[idx0].persistentContent = local_pProperties[idx0].persistentContent;
        }       
    }
```
As the pProperties is not copied over in the function at the moment. I also needed to modify terminator_CreateDisplayPlaneSurfaceKHR because it doesn’t map the pSurface using the unique_id_mapping system.

jon-lunarg commented:
AshleySmith:
Yes it looks like unique_objects needs an explicit_GetPhysicalDeviceDisplayPropertiesKHR like you said and the existing  explicit_GetDisplayPlaneSupportedDisplaysKHR() isn't quite right since it should reuse the existing mapping to unique_id rather than create a new one.

As for the explicit_CreateDisplayPlaneSurfaceKHR within unique_objects.cpp, what did you have to change? The surface mapping looks okay to me:

    if (VK_SUCCESS == result) {
        std::lock_guard<std::mutex> lock(global_lock);
// CODEGEN : file /home/jon/vk/vk/vk-layer-generate.py line #1594
        uint64_t unique_id = global_unique_id++;
        my_map_data->unique_id_mapping[unique_id] = reinterpret_cast<uint64_t &>(*pSurface);
        *pSurface = reinterpret_cast<VkSurfaceKHR&>(unique_id);
    }

jon-lunarg committed:
layers: Fix unique_objects creation of VkDisplayKHR

Make an explict_GetPhysicalDeviceDisplayProperties.

Change-Id: I39411d0dd97ca03dbf42d7e79b02243ef8988078

jon-lunarg commented:
Added patch to fix the unique_objects  GetPhysicalDeviceDisplayProperties and GetDisplayPlaneSupportedDisplays.

mark-lunarg commented on layers/unique_objects.h:
```diff
@@ -410,18 +410,56 @@ VkResult explicit_GetSwapchainImagesKHR(VkDevice device, VkSwapchainKHR swapchai
-                my_map_data->unique_id_mapping[unique_id] = reinterpret_cast<uint64_t &>(*pDisplays);
-                pDisplays[i] = reinterpret_cast<VkDisplayKHR&>(unique_id);
+                auto it = my_map_data->unique_id_mapping.find(reinterpret_cast<uint64_t> (pDisplays[i]));
+		assert (it !=  my_map_data->unique_id_mapping.end());
```
missingaspacehere.

pbantolas commented:
Jon:

That looks correct.

Although I think we should file a spec bug, it seems strange that you have to call GetPhysicalDeviceDisplayPropertiesKHR to initialise the displays. The spec is quite dubious on that aspect.

jon-lunarg commented:
Petros:

Go ahead and file a spec bug if you like for clarification.  I found the example usage in the spec does show calling GetPhysicalDeviceDisplayPropertiesKHR to get the display objects and then
GetDisplayPlaneSupportedDisplaysKHR to find which displays support  a given plane.

Ian:
I currently don't have time for adding all the validation checks to swapchain for this extension.  I did add TODOs in the code for the missing checks to be added later.

I fixed some compilation issues on Android and Windows  and pushing this series of patches to master branch.
